File: src/main/java/edu/school21/butterba/smartcalc/model/core/DepositCalcModel.java
package edu.school21.butterba.smartcalc.model.core;

import org.bytedeco.javacpp.Loader;
import org.bytedeco.javacpp.Pointer;
import org.bytedeco.javacpp.annotation.Name;
import org.bytedeco.javacpp.annotation.Namespace;
import org.bytedeco.javacpp.annotation.Platform;

@Platform(include = "deposit/deposit_calc_model.h", link = "smart_calc")
@Namespace("s21")
public class DepositCalcModel extends Pointer {

  static {
    Loader.load();
  }

  public DepositCalcModel() {
    allocate();
  }

  private native void allocate();

  @Name("CalculateDeposit")
  public native void calculateDeposit();

  @Name("SetInterestRate")
  public native void setInterestRate(final double interest_rate);

  @Name("SetTaxRate")
  public native void setTaxRate(final double tax_rate);

  @Name("SetMonthTerm")
  public native void setMonthTerm(final int month_term);

  @Name("SetIsCapitalized")
  public native void setIsCapitalized(final boolean is_capitalized);

  @Name("SetCapitalizationPeriodicity")
  public native void setCapitalizationPeriodicity(
      final int capitalization_periodicity);

  @Name("GetDepositAmount")
  public native double getDepositAmount();

  @Name("SetDepositAmount")
  public native void setDepositAmount(final double deposit_amount);

  @Name("GetInterest")
  public native double getInterest();

  @Name("GetTaxes")
  public native double getTaxes();

  @Name("AddReplenishment")
  public native void addReplenishment(final int month, final double value);

  @Name("AddWithdrawal")
  public native void addWithdrawal(final int month, final double value);

  @Name("Clear")
  public native void clear();
}

File: src/main/cpp/deposit/deposit_calc_model.cc
#include "deposit_calc_model.h"

#include "../smart_calc/smart_calc_model.h"

namespace s21 {

void DepositCalcModel::CalculateDeposit() {
  double capitalized = 0;
  interest_ = 0;
  for (int i = 0; i < month_term_; ++i) {
    deposit_amount_ += replenishment_[i];
    if (deposit_amount_ <= withdrawals_[i]) {
      throw std::out_of_range("Unable to withdraw more than deposit amount");
    }
    deposit_amount_ -= withdrawals_[i];
    double monthly_interest = deposit_amount_ * interest_rate_ / 12 / 100;
    if (is_capitalized_) {
      DepositCapitalization(&capitalized, monthly_interest, i);
    }
    interest_ += monthly_interest;
  }
  if (!is_capitalized_) {
    deposit_amount_ += interest_;
  }
  deposit_amount_ += capitalized;
  taxes_ = interest_ * tax_rate_ / 100;
  deposit_amount_ -= taxes_;
}

void DepositCalcModel::SetDepositAmount(double deposit_amount) {
  deposit_amount_ = deposit_amount;
}

void DepositCalcModel::SetInterestRate(double interest_rate) {
  interest_rate_ = interest_rate;
}

void DepositCalcModel::SetTaxRate(double tax_rate) { tax_rate_ = tax_rate; }

void DepositCalcModel::SetMonthTerm(int month_term) {
  month_term_ = month_term;
}

void DepositCalcModel::SetIsCapitalized(bool is_capitalized) {
  is_capitalized_ = is_capitalized;
}

void DepositCalcModel::SetCapitalizationPeriodicity(
    int capitalization_periodicity) {
  capitalization_periodicity_ = capitalization_periodicity;
}

double DepositCalcModel::GetDepositAmount() const { return deposit_amount_; }

double DepositCalcModel::GetInterest() const { return interest_; }

double DepositCalcModel::GetTaxes() const { return taxes_; }

void DepositCalcModel::AddReplenishment(int month, double value) {
  replenishment_[month] += value;
}

void DepositCalcModel::AddWithdrawal(int month, double value) {
  withdrawals_[month] += value;
}

void DepositCalcModel::Clear() {
  replenishment_.clear();
  withdrawals_.clear();
}

void DepositCalcModel::DepositCapitalization(double *capitalized,
                                             double monthly_interest,
                                             int month) {
  *capitalized += monthly_interest;
  if (!capitalization_periodicity_) {
    deposit_amount_ += *capitalized;
    *capitalized = 0;
  } else if (capitalization_periodicity_ == 1) {
    if (!(month % 4)) {
      deposit_amount_ += *capitalized;
      *capitalized = 0;
    }
  } else if (capitalization_periodicity_ == 2) {
    if (!(month % 6)) {
      deposit_amount_ += *capitalized;
      *capitalized = 0;
    }
  } else {
    if (!(month % 12)) {
      deposit_amount_ += *capitalized;
      *capitalized = 0;
    }
  }
}

} // namespace s21